src/precompile.c

```diff
@@ -67,41 +117,53 @@ JL_DLLEXPORT void jl_write_compiler_output(void)
             jl_precompile_toplevel_module = NULL;
     }
 
-    if (jl_options.incremental) {
-        if (jl_options.outputbc || jl_options.outputunoptbc)
-            jl_printf(JL_STDERR, "WARNING: incremental output to a .bc file is not implemented\n");
-        if (jl_options.outputasm)
-            jl_printf(JL_STDERR, "WARNING: incremental output to a .s file is not implemented\n");
-        if (jl_options.outputo) {
-            jl_printf(JL_STDERR, "WARNING: incremental output to a .o file is not implemented\n");
-        }
-    }
+    bool_t emit_native = jl_options.outputo || jl_options.outputbc || jl_options.outputunoptbc || jl_options.outputasm;
 
-    ios_t *s = jl_create_system_image(native_code, jl_options.incremental ? worklist : NULL);
+    bool_t emit_split = jl_options.outputji && emit_native;
 
-    if (jl_options.outputji) {
-        ios_t f;
-        if (ios_file(&f, jl_options.outputji, 1, 1, 1, 1) == NULL)
-            jl_errorf("cannot open system image file \"%s\" for writing", jl_options.outputji);
-        ios_write(&f, (const char*)s->buf, (size_t)s->size);
-        ios_close(&f);
-    }
+    ios_t *s = NULL;
+    ios_t *z = NULL;
+    int64_t srctextpos = 0 ;
+    jl_create_system_image(native_code, jl_options.incremental ? worklist : NULL, emit_split,
```

Comment:
It turns out we have an ordering issue:
- first we create `native_code = jl_precompile_worklist(worklist)`. The problem is, this only traverses modules in `worklist`. 
- later (in `jl_create_system_image`) we call `jl_prepare_serialization_data` and that creates `extext_methods` (the list of methods we own but which are methods of functions owned by other modules, and can only be found by traversing those modules) and `new_specializations` (new specializations of methods we don't own). We need to get both of these in `native_code`.

Is there any reason we can't delay the generation of `native_code` until we've set up both `extext_methods` and  `new_specializations`?

Reply:
Maybe we should merge `precompile.c` and `staticdata.c`?